Raise meaningful error when no serializable class defined for resource

smaximov opened the pull request:
This PR introduces a better error message than cryptic `undefined method 'new' for nil:NilClass` when no serializable class defined for a resource.

smaximov commented:
Just got similar "undefined method" error here: https://github.com/jsonapi-rb/jsonapi-serializable/blob/f7f1ac88ebb41b3b1b23d06f5bb672b9c10a5e36/lib/jsonapi/serializable/relationship/dsl.rb#L26-L28

Not sure how to handle it without code duplication.

smaximov commented:
@beauby, how about I move fetching serializable class to a separate class/module, so we can get meaningful error without code duplication?

beauby commented:
@smaximov Yeah, makes sense. It was all within a `ResourceBuilder` class before, but it had too complex an API. If you have a good name/architecture in mind, go ahead! :+1:

beauby commented:
I think for now we should go with a `JSONAPI::Serializable.class_for(object, hash)` module function.

smaximov commented:
> I think for now we should go with a JSONAPI::Serializable.class_for(object, hash) module function.

Implemented it in 0650192 and don't like it, reasons are boilerplate and duplication:

* As for now, there is no reason to have a helper method to return a serializable class because we always instantiate this class right away: [1](https://github.com/smaximov/jsonapi-serializable/blob/06501929260606fe1da7593e0647cb1821661e1e/lib/jsonapi/serializable/renderer.rb#L88-L89), [2](https://github.com/smaximov/jsonapi-serializable/blob/06501929260606fe1da7593e0647cb1821661e1e/lib/jsonapi/serializable/relationship/dsl.rb#L22-L23), [3](https://github.com/smaximov/jsonapi-serializable/blob/06501929260606fe1da7593e0647cb1821661e1e/lib/jsonapi/serializable/relationship/dsl.rb#L26-L27).

* And some more duplication: [1](https://github.com/smaximov/jsonapi-serializable/blob/06501929260606fe1da7593e0647cb1821661e1e/lib/jsonapi/serializable/renderer.rb#L76-L90), [2](https://github.com/smaximov/jsonapi-serializable/blob/06501929260606fe1da7593e0647cb1821661e1e/lib/jsonapi/serializable/relationship/dsl.rb#L18-L28).

> It was all within a ResourceBuilder class before, but it had too complex an API

I think having a separate class for this is a good idea. Aside from addressing the issues mentioned above, it helps testing resource building logic in isolation.

beauby commented:
Hi @smaximov – thanks for looking into this. The one thing I do not like about this `ResourceBuilder` pattern is that it creates a throw-away object for every relationship, hence my suggestion of a module function (previously, an instance of `ResourceBuilder` was passed around within exposures, but I don't really like that pattern either because it prevents calling `as_jsonapi` directly on a resource without knowledge of the internals).
How about having a `JSONAPI::Serializable.resource_for(object, inferrer)` then, which would instantiate the resource class? It would leave a tiny bit of duplication (namely the `if blah.respond_to?(:to_ary); ... ; else; ...; end`, but I think that's ok).

smaximov commented:
I agree with you that creating an instance of `J::S::ResourceBuilder` doesn't make much sense and it's not very good performance-wise.

> How about having a `JSONAPI::Serializable.resource_for(object, inferrer)` then, which would instantiate the resource class? 

Why not make `J::S::ResourceBuilder` a module then?

> It would leave a tiny bit of duplication (namely the `if blah.respond_to?(:to_ary); ... ; else; ...; end`, but I think that's ok).

I think it's easy to work around it.

beauby commented:
> Why not make J::S::ResourceBuilder a module then?

It's debatable, but I personally do not like modules/classes with only class methods. Moreover, I think building a resource/resources is central enough to the functionality of this gem that it makes sense to have those two methods on the main module, i.e.:
* `JSONAPI::Serializable.resource_for(...)`
* `JSONAPI::Serializable.resources_for(...)`

What do you think?

smaximov commented:
Personally I like having a separate module more (because it represents single concern), but I'm fine with your suggestion. So I'm moving resource building method to the main module?

beauby commented:
@smaximov Yes please 🎉

beauby commented on lib/jsonapi/serializable.rb:
```diff
@@ -1,3 +1,35 @@
+    class UndefinedSerializableClass < StandardError; end
+
+    # @api private
+    def self.resources_for(objects, options, inferrer)
```
Any specific rationale for not defining those with `module_function`? I do not feel strongly about this but just wondering.

smaximov replied:
AFAIK, `module_function` is meant to be used with modules that are mixed in classes, because it also creates private instance methods available to the instance methods of the class.

Since I don't think `JSONAPI::Serializable` will ever be included in any class, I didn't used `module_function`.

beauby replied:
Makes sense 👌

WaKeMaTTa replied:
@smaximov Thanks for sharing the explication of `module_function` :)

beauby commented on lib/jsonapi/serializable.rb:
```diff
@@ -1,3 +1,35 @@
+    require 'jsonapi/serializable/renderer'
+
+    # Error raised when there's no serializable class defined for resource.
+    class UndefinedSerializableClass < StandardError; end
```
Maybe `NameError` is more suitable as a superclass here?

beauby replied:
Sounds sensible 👌

beauby commented on lib/jsonapi/serializable/relationship/dsl.rb:
```diff
@@ -14,18 +14,7 @@ def data
-              @_class[resources.class.name.to_sym]
-                .new(@_exposures.merge(object: resources))
-            end
+            JSONAPI::Serializable.resources_for(yield, @_exposures, @_class)
```
It certainly does look better! ;)

beauby commented:
LGTM, merging. Thanks for your work @smaximov!

smaximov commented:
You're welcome :)

WaKeMaTTa commented:
@smaximov Thank you so much for this PR. You saved my day.

stevenharman commented:
@beauby Any chance of cutting a new release to get this out? I just tripped over the same thing.
